### PR TITLE
[ticket/14994] Refactor template->assign_block_var

### PR DESCRIPTION
**Pull request by javiexin**
Refactor assign_block_var to use the same block selection mechanism
as is used in alter_block_array.  This allows creating new blocks
at any position in the template structure, not only on the last block.
Allows selecting a block as outer[2].middle.

PHPBB3-14994

Checklist:

- [x] Correct branch: master for new features; 3.2.x, 3.1.x for fixes
- [x] Tests pass
- [x] Code follows coding guidelines: [master / 3.2.x](https://area51.phpbb.com/docs/master/coding-guidelines.html), [3.1.x](https://area51.phpbb.com/docs/31x/coding-guidelines.html)
- [x] Commit follows commit message [format](https://wiki.phpbb.com/Git#Commit_Messages)

Tracker ticket:

https://tracker.phpbb.com/browse/PHPBB3-14994

### CODE REVIEW

File: phpBB/phpbb/template/context.php [hunk 1]
```diff
@@ -179,70 +179,65 @@ public function &get_root_ref()
...
+			$vararray['S_FIRST_ROW'] = true;
+		}
+
+		// Assign S_BLOCK_NAME
+		$vararray['S_BLOCK_NAME'] = $name;
 
-			// Add a new iteration to this block with the variable assignments we were given.
-			$this->tpldata[$blockname][] = $vararray;
+		// Now the tricky part, we always assign S_LAST_ROW and remove the entry before
+		// This is much more clever than going through the complete template data on display (phew)
```
Comment (Elsensee) [inline phpBB/phpbb/template/context.php]: Is it? Is it really cleverer to set a value and unset the value before in each iteration?

Comment (javiexin) [inline phpBB/phpbb/template/context.php]: I don't know, but I guess so: this os the original unchanged code and comment...

Comment (javiexin) [inline phpBB/phpbb/template/context.php]: And by the way, it is in each nested block level, once per nesting level, so two/three times max.  That is the minimum possible if the block existed before.

Comment (Elsensee) [inline phpBB/phpbb/template/context.php]: Hm. Then I guess, it's fine.

**Comment by marc1706** [comment]
I'm missing a few tests that cover this change of behavior, e.g. tests that would fail before this PR and that do not fail now (as in setting a nested block for a previous block).

**Comment by javiexin** [comment]
@marc1706 Tests added per your request

**Comment by javiexin** [comment]
@marc1706 Anything else left to do here? Is this going forward?

**Comment by marc1706** [comment]
Replacing with refactored version without merge conflicts.

**Pull request closed without merging**